File: cpp/strategy.cpp
#include <iostream>

using namespace std;

class DrivingStrategy {
    public: 
        virtual void drive() = 0;
};

class FastDriving: public DrivingStrategy  {
    public: 
        void drive() {
            cout<<"driving fast"<<"\n";   
        }
};

class NormalDriving: public DrivingStrategy {
    public:
        void drive() {
            cout<<"driving normal"<<"\n";
        }
};

class Vehicle {
    DrivingStrategy* drivingStrategy;
    public:
    Vehicle(DrivingStrategy* drivingStrategy) {
        this->drivingStrategy = drivingStrategy;
    }

    void drive() {
        drivingStrategy->drive();
    }
};

class SportsVehicle: public Vehicle {
    public:
        SportsVehicle(): Vehicle(new FastDriving) {};
    
    void drive() {
        Vehicle::drive();
    }
};

class FamilyVehicle: public Vehicle {
    public:
        FamilyVehicle(): Vehicle(new NormalDriving) {};
    
    void drive() {
        Vehicle::drive();
    }
};

int main() {
    SportsVehicle sportsVehicle;
    FamilyVehicle familyVehicle;
    sportsVehicle.drive();
    familyVehicle.drive();
    return 0;
}

File: golang/strategy.go
package main

import "fmt"

type DrivingStrategy interface {
	Drive()
}

type FastDrivingStrategy struct {
}

func (f *FastDrivingStrategy) Drive() {
	fmt.Println("driving fast!")
}

type NormalDrivingStrategy struct {
}

func (n *NormalDrivingStrategy) Drive() {
	fmt.Println("normal driving!")
}

type Car interface {
	DriveCar()
}

type FamilyCar struct {
	strategy DrivingStrategy
}

func (f *FamilyCar) DriveCar() {
	f.strategy.Drive()
}

type RacingCar struct {
	strategy DrivingStrategy
}

func (r *RacingCar) DriveCar() {
	r.strategy.Drive()
}

func NewFamilyCar() Car {
	return &FamilyCar{strategy: &NormalDrivingStrategy{}}
}

func NewRacingCar() Car {
	return &RacingCar{strategy: &FastDrivingStrategy{}}
}

func main() {
	familyCar := NewFamilyCar()
	racingCar := NewRacingCar()

	familyCar.DriveCar()
	racingCar.DriveCar()
}
